weave/trace_server/orm.py

```diff
@@ -0,0 +1,621 @@
+    would use it as follows:
+
+    ```
+    pb = ParamBuilder()
+    param_name = pb.add_param("some_value")
+    query = f"SELECT * FROM some_table WHERE some_column = {{{param_name}:String}}"
+    parameters = pb.get_params()
+    # Execute the query with the parameters
+    ```
+
+    With queries that have many construction phases, it is recommended to use the
+    same ParamBuilder instance to ensure that the parameter names are unique across
+    the query.
+    """
+
+    def __init__(
+        self,
+        prefix: typing.Optional[str] = None,
+        database_type: DatabaseType = "clickhouse",
+    ):
+        global param_builder_count
+        param_builder_count += 1
+        self._params: typing.Dict[str, typing.Any] = {}
+        self._prefix = (prefix or f"pb_{param_builder_count}") + "_"
+        self._database_type = database_type
+
+    def add_param(self, param_value: typing.Any) -> str:
+        param_name = self._prefix + str(len(self._params))
+        self._params[param_name] = param_value
+        return param_name
+
+    def add(
+        self,
+        param_value: typing.Any,
+        param_name: typing.Optional[str] = None,
+        param_type: typing.Optional[str] = None,
+    ) -> str:
+        """Returns the placeholder for the target database type.
+
+        e.g. SQLite -> :limit
+        e.g. ClickHouse -> {limit:UInt64}
+        """
+        param_name = param_name or self._prefix + str(len(self._params))
+        self._params[param_name] = param_value
+        if self._database_type == "clickhouse":
+            ptype = param_type or _python_value_to_ch_type(param_value)
+            return f"{{{param_name}:{ptype}}}"
+        return ":" + param_name
+
+    def get_params(self) -> typing.Dict[str, typing.Any]:
+        return {**self._params}
+
+
+Value: TypeAlias = typing.Optional[
+    typing.Union[str, float, datetime.datetime, list[str], list[float]]
+]
+Row: TypeAlias = dict[str, Value]
+
+
+ColumnType = typing.Literal[
+    "string",
+    "datetime",
+    "json",  # Represented as string in ClickHouse
+]
+
+
+class Column:
+    # This is the name of the column from a user perspective.
+    name: str
+
+    # If specified, this is the name of the column in the database.
+    # Normally we just use name, but sometimes we have an internal convention like
+    # a "_dump" suffix that we don't want to expose in the API.
+    db_name: typing.Optional[str]
+    type: ColumnType
+    nullable: bool
+    # TODO: Description?
+    # TODO: Default?
+
+    def __init__(
+        self,
+        name: str,
+        type: ColumnType,
+        nullable: bool = False,
+        db_name: typing.Optional[str] = None,
+    ) -> None:
+        self.name = name
+        self.db_name = db_name
+        self.type = type
+        self.nullable = nullable
+
+    def dbname(self) -> str:
+        return self.db_name or self.name
+
+    def create_sql(self) -> str:
+        sql_type = "TEXT"
```

Comment:
should this be `self.type`?

Reply:
It seemed like in SQLite we were mapping all of the field types ORM currently supports (string, json, datetime) to a TEXT column, but this would need to be updated with an actual mapping to support other column types.